library/core/src/iter/adapters/zip.rs

```diff
@@ -171,26 +246,109 @@ macro_rules! zip_impl_general_defaults {
             // and doesn’t call `next_back` too often, so this implementation is safe in
             // the `TrustedRandomAccessNoCoerce` specialization
 
-            let a_sz = self.a.len();
-            let b_sz = self.b.len();
-            if a_sz != b_sz {
-                // Adjust a, b to equal length
-                if a_sz > b_sz {
-                    for _ in 0..a_sz - b_sz {
-                        self.a.next_back();
-                    }
-                } else {
-                    for _ in 0..b_sz - a_sz {
-                        self.b.next_back();
-                    }
-                }
-            }
+            self.adjust_back();
             match (self.a.next_back(), self.b.next_back()) {
                 (Some(x), Some(y)) => Some((x, y)),
                 (None, None) => None,
                 _ => unreachable!(),
             }
         }
+
+        #[inline]
+        default fn fold<T, F>(self, init: T, mut f: F) -> T
+        where
+            F: FnMut(T, Self::Item) -> T,
+        {
+            let mut a = self.a;
+            let mut b = self.b;
+
+            let acc = a.try_fold(init, move |acc, x| match b.next() {
+                Some(y) => Ok(f(acc, (x, y))),
+                None => Err(acc),
+            });
+
+            match acc {
+                Ok(exhausted_a) => exhausted_a,
+                Err(exhausted_b) => exhausted_b,
+            }
+        }
```

Comment:
When a `fold` implementation can't be implemented on top of an inner `fold`, it's common in libcore to delegate to `self`'s `try_fold` instead in order to avoid some duplicate logic. See e.g. [`Take::fold`](https://github.com/rust-lang/rust/blob/master/library/core/src/iter/adapters/take.rs#L101-L113). `TakeWhile`, `MapWhile`, and `Scan` do the same thing. You could see if it affects benchmark results in any way, but I'd hope that it won't.

SIde note: closures in the implementation of iterator adapters are typically not written outright, but returned from a nested function, see #62429.